src/Workspaces/CSharp/Portable/Editing/CSharpImportAdder.cs

```diff
@@ -142,91 +129,137 @@ private void AddImportedMembers(ImmutableArray<INamespaceSymbol> namespaceSymbol
-        }
+            using var _ = ArrayBuilder<SyntaxNode>.GetInstance(out var nodes);
 
-        public override void VisitParenthesizedLambdaExpression(ParenthesizedLambdaExpressionSyntax node)
-        {
-            var previousInAnonymousMethod = _inAnonymousMethod;
-            _inAnonymousMethod = true;
-            base.VisitParenthesizedLambdaExpression(node);
-            _inAnonymousMethod = previousInAnonymousMethod;
+            CollectInfoFromContainer(container, nodes, out var containsAnonymousMethods);
+
+            await ProducerConsumer<INamespaceSymbol>.RunParallelAsync(
+                source: nodes,
+                produceItems: static (node, onItemsFound, args, cancellationToken) =>
+                {
+                    var (self, containsAnonymousMethods, _) = args;
+                    if (node is SimpleNameSyntax nameSyntaxNode)
+                        self.ProduceConflicts(nameSyntaxNode, onItemsFound, cancellationToken);
+                    else if (node is MemberAccessExpressionSyntax memberAccessExpressionNode)
+                        self.ProduceConflicts(memberAccessExpressionNode, containsAnonymousMethods, onItemsFound, cancellationToken);
+                    else
+                        throw ExceptionUtilities.Unreachable();
+
+                    return Task.CompletedTask;
+                },
+                consumeItems: static async (items, args, cancellationToken) =>
+                {
+                    var (_, _, conflicts) = args;
+                    await foreach (var conflict in items)
+                        conflicts.Add(conflict);
+                },
+                args: (self: this, containsAnonymousMethods, conflicts),
+                cancellationToken).ConfigureAwait(false);
         }
 
-        public override void VisitAnonymousMethodExpression(AnonymousMethodExpressionSyntax node)
+        private void CollectInfoFromContainer(SyntaxNode container, ArrayBuilder<SyntaxNode> nodes, out bool containsAnonymousMethods)
         {
-            var previousInAnonymousMethod = _inAnonymousMethod;
-            _inAnonymousMethod = true;
-            base.VisitAnonymousMethodExpression(node);
-            _inAnonymousMethod = previousInAnonymousMethod;
+            containsAnonymousMethods = false;
+
+            foreach (var node in container.DescendantNodesAndSelf())
+            {
+                switch (node.Kind())
+                {
+                    case SyntaxKind.IdentifierName:
+                    case SyntaxKind.GenericName:
+                        if (IsPotentialConflictWithImportedType((SimpleNameSyntax)node))
+                            nodes.Add(node);
+                        break;
+                    case SyntaxKind.SimpleMemberAccessExpression:
+                    case SyntaxKind.PointerMemberAccessExpression:
+                        if (IsPotentialConflictWithImportedExtensionMethod((MemberAccessExpressionSyntax)node))
```

Comment:
fwiw, i don't think the pointer case can happen with imports and extensions.  but i don't mind you preserving this.